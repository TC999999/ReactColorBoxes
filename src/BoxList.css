.Boxes {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.box-list-div {
  display: flex;
  flex-direction: column;
  align-items: center;
}
