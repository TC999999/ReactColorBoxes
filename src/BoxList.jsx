import { useState } from "react";
import { v4 as uuid } from "uuid";
import Box from "./Box.jsx";
import NewBoxForm from "./NewBoxForm.jsx";
import "./BoxList.css";

const BoxList = () => {
  const initialState = [];
  const [boxes, setBoxes] = useState(initialState);

  const addBox = (newBox) => {
    setBoxes((boxes) => [...boxes, { ...newBox, id: uuid() }]);
  };

  const delBox = (e) => {
    let boxesCopy = boxes.map((b) => {
      return { ...b };
    });
    let id = e.target.parentElement.id;
    let filteredBoxesCopy = boxesCopy.filter((b) => {
      return b.id !== id;
    });
    setBoxes(filteredBoxesCopy);
  };

  return (
    <div className="Boxes">
      <h3>Colored Boxes</h3>
      <NewBoxForm addBox={addBox} />
      <div className="box-list-div">
        {boxes.map(({ id, color, height, width }) => {
          return (
            <Box
              id={id}
              key={id}
              color={color}
              height={height}
              width={width}
              delBox={delBox}
            />
          );
        })}
      </div>
    </div>
  );
};

export default BoxList;
