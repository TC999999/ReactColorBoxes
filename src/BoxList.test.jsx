import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList.jsx";

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("makes a box when the form is submitted", function () {
  const { queryByText, getByLabelText, container } = render(<BoxList />);
  const colorInput = getByLabelText("Color");
  const heightInput = getByLabelText("Height");
  const widthInput = getByLabelText("Width");

  const btn = queryByText("Make a Box!");

  expect(
    container.querySelector('div[class="box-div"]')
  ).not.toBeInTheDocument();
  expect(container.querySelector('div[class="box"]')).not.toBeInTheDocument();

  fireEvent.change(colorInput, { target: { value: "purple" } });
  fireEvent.change(heightInput, { target: { value: "100" } });
  fireEvent.change(widthInput, { target: { value: "100" } });
  fireEvent.submit(btn);

  expect(container.querySelector('div[class="box-div"]')).toBeInTheDocument();
  const box = container.querySelector('div[class="box"]');
  expect(box).toBeInTheDocument();
  expect(box.style["background-color"]).toBe("purple");
  expect(box.style["height"]).toBe("100px");
  expect(box.style["width"]).toBe("100px");
});

it("deletes a box when the 'X' button is clicked", function () {
  const { queryByText, getByLabelText, container } = render(<BoxList />);
  const colorInput = getByLabelText("Color");
  const heightInput = getByLabelText("Height");
  const widthInput = getByLabelText("Width");
  const btn = queryByText("Make a Box!");
  fireEvent.change(colorInput, { target: { value: "purple" } });
  fireEvent.change(heightInput, { target: { value: "100" } });
  fireEvent.change(widthInput, { target: { value: "100" } });
  fireEvent.submit(btn);
  expect(container.querySelector('div[class="box"]')).toBeInTheDocument();

  const delBtn = queryByText("X");
  fireEvent.click(delBtn);
  expect(container.querySelector('div[class="box"]')).not.toBeInTheDocument();
});
