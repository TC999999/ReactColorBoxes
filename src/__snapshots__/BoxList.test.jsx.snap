// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Boxes"
  >
    <h3>
      Colored Boxes
    </h3>
    <form>
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        placeholder="color of box"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        placeholder="height of box in px"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        placeholder="width of box in px"
        type="text"
        value=""
      />
      <button>
        Make a Box!
      </button>
    </form>
    <div
      class="box-list-div"
    />
  </div>
</DocumentFragment>
`;
